### ISSUE TITLE
Exponentially slow import and export performance

### ISSUE THREAD

**Issue by n2k3s2it** [KhronosGroup/glTF-Blender-IO#1078]
**Describe the bug**
Hello, it seems I'm consistently running into performance issues during import (glTF) and export (glb, no Draco).

**Test results in Blender 2.80.75 (comes with io_scene_gltf2 0.9.36):**
Import (batched):
- 209+83+125+138 separate glTF/bin pairs into 4 separate blend files: **8:03 minutes**
- 555 separate glTF/bin pairs into 1 blend file: **34:33 minutes**

Export (batched):
- 209+83+125+138 groups exported from 4 separate blend files: **11:47 minutes**
- 555 groups of objects exported from 1 blend file: **52:15 minutes**

**Test results in Blender 2.83.16 bcon3 (comes with io_scene_gltf2 1.2.75):**
Import (batched):
- 209+83+125+138 separate glTF/bin pairs into 4 separate blend files: **3:16 minutes**
- 555 separate glTF/bin pairs into 1 blend file: **15:20 minutes**

Export (batched):
- 209+83+125+138 groups exported from 4 separate blend files: **12:43 minutes**
- 555 groups of objects exported from 1 blend file: **54:53 minutes**

While import results in Blender 2.83.16 are faster, both import and export scenario's are still at least 4x slower when performed in 1 blend file.

My conclusion is the more objects exist in the blender scene where additional glTF's are imported into, or exported from the longer it takes to import the same glTF file or export the same data into a glb file. This is confirmed by me manually importing 1 glTF/bin pair into an empty scene it takes 0.38 seconds, and when doing this while having a blend file open that already has all glTF/bin pairs imported, it takes 13.39 seconds for the same glTF/bin pair.

**Expected behavior**
In both scenario's for import and export I expect it to take the same amount of time, as it is processing the same data.

**To Reproduce Import scenario**
1. Start Blender 2.80.75
3. Open a new empty scene
2. Import `0_1_1.gltf` (from [these glTF files](https://github.com/n2k3s2it/n2k3s2it.github.io/blob/master/blender/glTF%20files.zip))
4. It takes 0.25 seconds
5. Open the [Blend file](https://github.com/n2k3s2it/n2k3s2it.github.io/blob/master/blender/Blend%20file.zip)
6. Import `0_1_1.gltf` into this scene
7. It now takes 7.35 seconds

**To Reproduce Export scenario**
1. export all objects imported into an empty scene it takes 0.67 seconds.
2. export all objects in collection 0_1_1 from the provided blend file it takes 7.3 seconds.

**Tested on:**
OS:
- Windows 10 Pro 1803 (Intel Core i5-8600, 6C-6T, Base: 3.1 GHz, Max Turbo: 4.3 GHz)
- CentOS 7.7.1908 (Intel Xeon E5640, 4C-4T, Base: 2.66 GHz, Max Turbo: 2.93 GHz)

**Additional context**
Batched means:
- For import: using a custom python script resolving a directory containing glTF files and importing them in a simple `for` loop. The only non default setting that is used is `import_shading='FLAT'`.
- For export: using a custom python script grabbing collection contents in a simple `for` loop and exporting them into glb files. The following non default settings are used: `export_selected=True`, `export_materials=False`, `export_texcoords=True`, `export_extras=True`, `export_apply=True`

Times listed above are from Window 10 machine, but CentOS machine shows the same behavior. I've not tested blender 2.83.16 bcon3 on CentOS.

In all test scenario's the exact same data is being imported and exported. glTF/bin pairs only contain Nodes (referenced in 1 Scene), Meshes and their respective Accessors and BufferViews, along with materials containing pbrMetallicRoughness properties but there are no textures included. All glTF/bin pairs are valid according to the [glTF Validator](http://github.khronos.org/glTF-Validator/) and are version 2.0.

I have more data sets that are similar as tested here, they show the same behavior. Including the behavior when the data set is smaller or bigger, the import and export time decrease or increase respectively. I have data sets that take 3 hours to import and 1.5 hours to export.

This issue might be related to https://github.com/KhronosGroup/glTF-Blender-IO/issues/76

**Comment by julienduroure** [comment]
Hello,

Please provide some test cases. I know that you are not allowed to share the files you used, but without test files, we want analyse what the issue is.
Can you please create some similar test cases that you can share ? (I mean with same number of objects / same textures, etc...)

**Comment by scurest** [comment]
If I take the level.gltf file from #862 and import it repeatedly, each import takes about 1 second longer than the last.

Running with `--debug-value 102` to use garyo's profiler, the extra time looks like its coming from scripts/modules/bpy/ops.py:_view_layer_update. I stuck a `traceback.print_stack()` in that function and found it gets called when we do `bpy.ops.obect.mode_set(mode='OBJECT')`.

We do that mainly from the code for creating bones in armatures.

**Comment by n2k3s2it** [comment]
Hey, I was able to reproduce this using these [glTF files](https://github.com/n2k3s2it/n2k3s2it.github.io/blob/master/blender/glTF%20files.zip), or see the [blend file](https://github.com/n2k3s2it/n2k3s2it.github.io/blob/master/blender/Blend%20file.zip) where all glTF's are already imported.

If I import `0_1_1.gltf` in an empty scene it takes 0.25 seconds.
If I import `0_1_1.gltf` in the provided blend file, it takes 7.35 seconds.

**Comment by scurest** [comment]
@n2k3s2it I can't reproduce. Can you run blender like `blender --debug-value 102` and do the import into the provided blend file? Profiling information should be printed to the console. Paste that here.

**Comment by n2k3s2it** [comment]
I run like this

1. open cmd with admin
2. navigate to blender install folder `C:\Program Files\Blender Foundation\Blender 2.80`
3. type `blender --debug-value 102` and hit enter
4. blender starts and I reproduce the issue

Results:
```
Microsoft Windows [Version 10.0.17134.648]
(c) 2018 Microsoft Corporation. All rights reserved.

C:\Program Files\Blender Foundation\Blender 2.80>blender --debug-value 102
Read prefs: C:\Users\<redacted>\AppData\Roaming\Blender Foundation\Blender\2.80\config\userpref.blend
AL lib: (EE) UpdateDeviceParams: Failed to set 48000hz, got 44100hz instead
found bundled python: C:\Program Files\Blender Foundation\Blender 2.80\2.80\python
Read blend: D:\Files\blenderDemo\glTF_files_imported.blend
2020-05-29 10:18:26,914 CRITICAL Data are loaded, start creating Blender stuff
2020-05-29 10:18:34,912 CRITICAL glTF import finished in 8.00s
```

it seems nothing is being profiled though?

**Comment by scurest** [comment]
Use your 2.83.

**Comment by n2k3s2it** [comment]
Ok my bad testing 2.80 as this is our production version.

In 2.83 the results seem negligible (but this might be within margin of error)
- import in empty scene 0.02, see results: https://pastebin.com/raw/2pJeP9a7
- import in provided blend file: 0.07, see results: https://pastebin.com/raw/Aa07UsJW

**Comment by scurest** [comment]
Yeah, those are the times I got.

The `BlenderMesh.create` function has moved up the profile a bit.

Can you reproduce the 0.38 seconds / 13.39 seconds from the OP in 2.83? And post the profiles for that if you can?

> in 2.83 the export results are still a big difference (but no profiling info was shown)

Uh-huh. There's no profiler in the exporter.

**Comment by n2k3s2it** [comment]
> This is confirmed by me manually importing 1 glTF/bin pair into an empty scene it takes 0.38 seconds, and when doing this while having a blend file open that already has all glTF/bin pairs imported, it takes 13.39 seconds for the same glTF/bin pair.

This was done in 2.80 not in 2.83.

See the results below for the data set as tested in OP.

In 2.83 the import results still seem negligible (again within margin of error)

- import in empty scene 0.16, see results: https://pastebin.com/raw/9MFcvViW
- import in blend file: 0.20, see results: https://pastebin.com/raw/MTZSpKVB

in 2.83 the export results are again a big difference

- export all objects imported into an empty scene it takes 1.9 seconds.
- export all objects in collection 0_0_0 from the blend file it takes 23.60 seconds.

Import stats might be negligible for 1 glTF, but apparently it stacks when you do it 555 times :)

**Comment by julienduroure** [comment]
Hello,
Let's talk only to 2.83 (and master) version, as this is the only version that we are maintaining actively.

If I understand correctly the discussion, there is no issue regarding importing.

There is a performance issue regarding exporter:
* No issue when exporting all objects
* Issue when exporting only some objects == > How do you manage the exported objects? By using the "limit to selection" option?

Don't hesitate to tell me if I wrongly understand :)

**Comment by julienduroure** [comment]
2.5 months without any answer. Maybe we can close this ticket, or please comment to help us identify where you have some issue

**Comment by jcdeblok** [comment]
For what it's worth I found that GLB exported from CLO (3d fashion/cloth software) imports extremely slow as described in various bug reports here.  But when loaded into window's default 3d Viewer and exported as-is to a new GLB file it will load that file near instantly in blender. I can't share the model (client's next-season fashion items ) but if you have a slow loading GLB it's worth loading/exporting  it though windows's native 3D Viewer.

I did take a look at the data structures on the GLB's internally, there is some difference in how it handles textures. Clo export has 3 images, 3 textures and 8 materials.  3dViewer's export has 3 images, 24 textures and 8 materials, it has made unique texture nodes per material that do point all to the same image data. So it's not extra texture data, just different pointers to the same thing instead of sharing a pointer among materials.

Maybe this will help debugging/create a workaround for someone.
